indigo_api/migrations/0125_migrate_commencements.py

```diff
@@ -0,0 +1,84 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.11.27 on 2020-01-09 14:59
+from __future__ import unicode_literals
+
+from django.db import migrations
+
+from indigo_api.models import Act
+from indigo.plugins import plugins
+
+
+def migrate_commencements(apps, schema_editor):
+    """ Copy commencement details from works to Commencement objects
+    """
+    Work = apps.get_model("indigo_api", "Work")
+    Document = apps.get_model("indigo_api", "Document")
+    Commencement = apps.get_model("indigo_api", "Commencement")
+    db_alias = schema_editor.connection.alias
+
+    def for_document(topic, document):
+        country_code = document.work.country.country.iso.lower()
+        locality = document.work.locality.code if document.work.locality else None
+        language_code = document.language.language.iso_639_2B
+        return plugins.for_locale(topic, country=country_code, locality=locality, language=language_code)
+
+    def table_of_contents(expression):
+        builder = for_document('toc', expression)
+        expression.doc = Act(expression.document_xml)
+        expression.django_language = expression.language.language.iso_639_1
+        return builder.table_of_contents_for_document(expression)
+
+    def get_all_provisions(work):
+        toc = None
+        ids = []
+
+        first_expression = Document.objects.filter(work=work, deleted=False).order_by('expression_date').first()
+        if first_expression:
+            toc = table_of_contents(first_expression)
+
+        def add_ids(toc):
+            for e in toc:
+                if e.id:
+                    ids.append(e.id)
+                if e.children and e.component == 'main':
+                    add_ids(e.children)
+
+        if toc:
+            add_ids(toc)
+        return ids
+
+    for w in Work.objects.using(db_alias).all():
+        if w.commencing_work or w.commencement_date:
+            all_provisions = get_all_provisions(w)
+            commencement = Commencement(
+                commenced_work=w,
+                commencing_work=w.commencing_work,
+                date=w.commencement_date,
+                main=True,
+                all_provisions=True,
+                provisions_list=all_provisions,
```

Comment:
If we're introducing logic somewhere else that will "expand" the all provisions list, we could skip doing it here in the migration and simply delegate to that. Just a thought for now.

Reply:
yeah this should actually come out, I realised recently getting the first PiT for the ToC won't necessarily be right because we don't always have a document for the first PiT (e.g. starting with a consolidation). 

Can we make it so that only works with multiple commencements need `provisions` for their Commencement objects, and that these must be entered manually (at least to start)?